### PR TITLE
RFR: Update file structure doc post major restructure

### PR DESCRIPTION
**Pull request by Centrus007**

### CODE REVIEW

**Comment by SimonBiggs** [comment]
This one's going to be an undertaking...

**Comment by Centrus007** [comment]
How's this for a start?

File: docs/developer/file-structure.rst [hunk 1]
```diff
@@ -285,6 +285,8 @@ levelling. PyMedPhys' automated test suite includes a Python package called
 For a further, in-depth explanation of the philosophy behind levelling
 dependencies, see the :ref:`john-lakos` section.
 
+|
```
Comment (SimonBiggs) [inline docs/developer/file-structure.rst]: In the CSS I have now globally formatted greater section breaks by default:

https://github.com/pymedphys/pymedphys/blob/master/docs/_static/theme_overrides.css#L104-L106

I can make these section breaks even larger if you like?

File: docs/developer/file-structure.rst [hunk 2]
```diff
@@ -10,8 +10,8 @@ The PyMedPhys repository has the following general structure:
 
    pymedphys/
    |   README.rst
-   |   LICENSE
-   |   Apache-2.0
+   |   LICENSE-AGPL-3.0-or-later
+   |   LICENSE-Apache-2.0
    |   changelog.md
    |   setup.py
    |   layers.yml
```
Comment (SimonBiggs) [inline docs/developer/file-structure.rst]: We can leave it there for now, but after the big split we are no longer running layer linter. So until that gets sorted should we just leave the layer description out?

File: docs/developer/file-structure.rst [hunk 3]
```diff
@@ -75,112 +76,175 @@ omitted from this page in the interest of brevity.
...
-                     Code within this constitutes "PyMedPhys proper". In theory
-                     (though not yet in practice), this code has been tested
-                     and documented. Changes to code in this folder are tracked
-                     in ``changelog.md``. See :ref:`source-code` section below
-                     for further details, especially on structure.
+                 permitting both remote and collaborative raw development.
+
+:``packages/``: The PyMedPhys source code library is separated into a set of
+                **subpackages**, from which the main PyMedPhys package draws.
+                Users are strictly able to install any of these subpackages,
```
Comment (SimonBiggs) [inline docs/developer/file-structure.rst]:
```suggestion
                Users are able to install these subpackages standalone,
```

File: docs/developer/file-structure.rst [hunk 4]
```diff
@@ -75,112 +76,175 @@ omitted from this page in the interest of brevity.
...
+                but almost all will instead install the main PyMedPhys
+                package. The instructions given in `Installation`_ apply to the
+                main PyMedPhys package. Subpackages are not really intended for
+                user installation, but rather to help manage (the very complex)
```
Comment (SimonBiggs) [inline docs/developer/file-structure.rst]:
```suggestion
                user installation, but rather to provide an optional way to deploy applications with minimal dependencies when needed.
```

File: docs/developer/file-structure.rst [hunk 5]
```diff
@@ -75,112 +76,175 @@ omitted from this page in the interest of brevity.
...
-helps to prevent circular code dependencies. See diagram below:
+Almost all users will access the PyMedPhys library of source code via the
+main `pymedphys` package (*pymedphys/src/pymedphys/*). However no library code
+actually exists within *pymedphys/src/pymedphys/*. Instead, library code is
+contained within *pymedphys/packages/* and redirected through
+*pymedphys/src/pymedphys/* via a set of python imports.
+
+Within *pymedphys/packages/*, code is organised into a set of subpackages,
+such as ``pymedphys_analysis`` and ``pymedphys_dicom``. From there, each
+subpackage contains a directory named `src/<package_name>`. Within each
```
Comment (SimonBiggs) [inline docs/developer/file-structure.rst]:
```suggestion
subpackage contains a directory named ``src/<package_name>``. Within each
```

Do we want italics here or code blocks?

File: docs/developer/file-structure.rst [hunk 6]
```diff
@@ -192,33 +256,35 @@ level should only depend on code in files of lower-numbered levels. Code should
...
-``dicom``, such as ``d4a.py``, but no file within ``dicom`` is allowed to
-import from any file in ``gamma``.
-
-We are able to programatically check for any improper file levelling.
-PyMedPhys' automated test suite includes a Python package called
+the same level) or ``g3a.py`` (which is in a higher-numbered level).
+
+*This philosophy applies for modules (categories within subpackages) as well:*
+Each module has an assigned level. A module's level is flexible; it can be
+adjusted as needed. Modules are assigned levels in the file ``layers.yml``.
```
Comment (SimonBiggs) [inline docs/developer/file-structure.rst]: See other comment regarding layers.

File: docs/developer/file-structure.rst [hunk 7]
```diff
@@ -192,33 +256,35 @@ level should only depend on code in files of lower-numbered levels. Code should
...
+writing, ``mudensity`` is a level 2 module, and ``gamma`` is a ``level 3``
+module. This means that any file within ``gamma``, such as ``g1a.py``, is free
+to import from any file within ``mudensity``, such as ``m2a.py``, regardless of
+that file's level within its own module. However, no file within ``mudensity``
+is allowed to import from any file in ``gamma``. Note that a module's level is
+unaffected by which subpackage/s it is in.
+
+Thankfully, we are able to programatically check for any improper file
+levelling. PyMedPhys' automated test suite includes a Python package called
 ``layer-linter``, which does just that!
```
Comment (SimonBiggs) [inline docs/developer/file-structure.rst]: See other comment regarding layers.

File: docs/developer/file-structure.rst (hunk 5 above)
Comment (Centrus007) [reply docs/developer/file-structure.rst]: I struggled to settle on a general convention, to be honest. I started using italics for paths and code blocks for module names (and code, of course), but then broke that rule further down when discussing levelling.

Maybe I'll just lock in code blocks for paths as well.

**Comment by SimonBiggs** [comment]
If it helps define a convention, code blocks for paths is certainly a
preference of mine. (But I wouldn't fight on a hill for it)

**Comment by Centrus007** [comment]
> Thanks @Centrus007 big smiles here, this is awesome. Hugely appreciated!

No worries. Does it read okay? I kinda felt like it wasn't too bad of a start but could do with a rework one day (if anyone ever reads it, ha!)

**Comment by Centrus007** [comment]
> If it helps define a convention, code blocks for paths is certainly a preference of mine. (But I wouldn't fight on a hill for it)

Let's lock it in :)

**Comment by SimonBiggs** [comment]
> > Thanks @Centrus007 big smiles here, this is awesome. Hugely appreciated!
> 
> No worries. Does it read okay? I kinda felt like it wasn't too bad of a start but could do with a rework one day (if anyone ever reads it, ha!)

It certainly does read well. Definitely a good start. I know that one of the first things @msobolewski does when he tries to dive into a project is he wants to understand where everything is... I don't always get it, but it really helps him understand what's going on. ... so for a certain kind of person I envisage it will be invaluable.

**Pull request merged**